### PR TITLE
[FIX] fix pipeshard parallel without apply gradient

### PR DESCRIPTION
**Pull request by ZYHowell**

### CODE REVIEW

File: alpa/pipeline_parallel/apply_grad.py [hunk 1]
```diff
@@ -302,29 +310,33 @@ def _get_apply_grad_outvar_constraints(jax_pipeline_stages, stage_to_mesh,
...
                                       stage_to_mesh, mask)
+
+    # 2. Add compute mean and slice apply-grad stages
     # FIXME (zhuohan): get_mean only works when we use jax.mean to
     #                  calculate loss. It will fail if we use sum.
-    apply_grad_jaxpr, global_outvars = apply_grad_get_mean(
-        apply_grad_jaxpr, gradients, gensym_func, num_micro_batches,
-        global_outvars)
+    # apply_grad_jaxpr, global_outvars = apply_grad_get_mean(
+    #     apply_grad_jaxpr, gradients, gensym_func, num_micro_batches,
```
Comment (zhuohan123) [inline alpa/pipeline_parallel/apply_grad.py:332]: Remove this comment and move the FIXME above to around line 260

**Review by zhuohan123 — CHANGES_REQUESTED** [review]

**Review by zhuohan123 — APPROVED** [review]
LGTM

File: alpa/pipeline_parallel/apply_grad.py [hunk 2]
```diff
@@ -302,29 +316,28 @@ def _get_apply_grad_outvar_constraints(jax_pipeline_stages, stage_to_mesh,
...
+                           num_meshes, global_invars, global_outvars,
+                           donated_invars):
     """Slice apply_grad jaxpr into stages and assign them to the correspondig meshes."""
     # TODO(yonghao): the condition of creating RDA variable should be extended.
 
     # Process apply gradient:
     # 1. change invars of apply grad to outvars of accumulate grad
     gradients = [g for g in barrier.outvars if not isinstance(g, DropVar)]
+    assert len(gradients) == len(barrier.invars)
     mask = {
```
Comment (zhuohan123) [inline alpa/pipeline_parallel/apply_grad.py:328]: Not related to this PR. Can you change the name of this variable from `mask` to something meaningful? For example `barrier_out2in`
